support/agent/sim.go

```diff
@@ -0,0 +1,268 @@
+)
+
+type Sim struct {
+	Config        SimConfig
+	Accounts      []address.Address
+	Miners        []*MinerAgent
+	v             *vm.VM
+	rnd           *rand.Rand
+	statsByMethod map[vm.MethodKey]*vm.CallStats
+}
+
+type VMState interface {
+	GetEpoch() abi.ChainEpoch
+	GetState(addr address.Address, out cbor.Unmarshaler) error
+	Store() adt.Store
+}
+
+type SimConfig struct {
+	AccountCount           int
+	AccountInitialBalance  abi.TokenAmount
+	Seed                   int64
+	CreateMinerProbability float32
+}
+
+type ReturnHandler func(v VMState, msg Message, ret cbor.Marshaler) error
+
+type Message struct {
+	From          address.Address
+	To            address.Address
+	Value         abi.TokenAmount
+	Method        abi.MethodNum
+	Params        interface{}
+	ReturnHandler ReturnHandler
+}
+
+type MinerPowerTable struct {
+	addr    address.Address
+	qaPower abi.StoragePower
+}
+
+type PowerTable struct {
+	blockReward  abi.TokenAmount
+	totalQAPower abi.StoragePower
+	minerPower   []MinerPowerTable
+}
+
+func NewSim(ctx context.Context, t require.TestingT, store adt.Store, config SimConfig) *Sim {
+	v := vm.NewCustomStoreVMWithSingletons(ctx, store, t)
+	return &Sim{
+		Config:   config,
+		Accounts: vm.CreateAccounts(ctx, t, v, config.AccountCount, config.AccountInitialBalance, config.Seed),
+		Miners:   []*MinerAgent{},
+		v:        v,
+		rnd:      rand.New(rand.NewSource(config.Seed)),
+	}
+}
+
+func (s *Sim) Tick() error {
+	var err error
+	var blockMessages []Message
+
+	// compute power table before state transition to create block rewards at the end
+	powerTable, err := ComputePowerTable(s.v, s.Miners)
+	if err != nil {
+		return err
+	}
+
+	// add all miner messages
+	for _, miner := range s.Miners {
+		msgs, err := miner.Tick(s.v)
+		if err != nil {
+			return err
+		}
+
+		blockMessages = append(blockMessages, msgs...)
+	}
+
+	// add at most 1 miner per epoch.
+	if len(s.Miners) < len(s.Accounts) && s.rnd.Float32() < s.Config.CreateMinerProbability {
+		addr := s.Accounts[len(s.Miners)]
+		blockMessages = append(blockMessages, s.createMiner(addr, MinerAgentConfig{
+			PrecommitRate:   2.5,
+			ProofType:       abi.RegisteredSealProof_StackedDrg32GiBV1,
+			StartingBalance: s.Config.AccountInitialBalance, // miner gets all account funds
+		}))
+	}
+
+	// shuffle messages
+	s.rnd.Shuffle(len(blockMessages), func(i, j int) {
+		blockMessages[i], blockMessages[j] = blockMessages[j], blockMessages[i]
+	})
+
+	// run messages
+	for _, msg := range blockMessages {
+		ret, code := s.v.ApplyMessage(msg.From, msg.To, msg.Value, msg.Method, msg.Params)
+
+		// for now, assume everything should work
+		if code != exitcode.Ok {
+			return errors.Errorf("exitcode %d: message failed: %v\n%s\n", code, msg, strings.Join(s.v.GetLogs(), "\n"))
+		}
+
+		if msg.ReturnHandler != nil {
+			if err := msg.ReturnHandler(s.v, msg, ret); err != nil {
+				return err
+			}
+		}
+	}
+
+	// apply block rewards
```

Comment:
While not impossible to see on mainnet (i.e. in the case of all tipset blocks containing identical messages) this sequencing is unlikely so will move simulations further away from the expected behavior of [interleaving rewards](https://github.com/filecoin-project/lotus/blob/2a294f048ac211d999ea6b044fe4412963901fe7/chain/stmgr/stmgr.go#L353) between blocks.  It is probably not a very important difference.

Reply:
Hmm. I thought all rewards came at the end of the tipset. I guess if this is the case, shuffling the rewards in with the other message might actually be more realistic (except there could be a reward at zero and it is unlikely that the last message will be a reward).

Reply:
Please just leave a comment highlighting this difference from the real chain, so it's transparent if anyone runs into it later.